File: 프로그래머스/lv1/131705. 삼총사/삼총사.js
function solution(number) {
    let check = Array.from({length:number.length},()=>0);
    let answer = 0;
    const DFS = (start,sum,cnt)=>{
        if(cnt === 3 && sum === 0){
           answer++
        }
        
        for(let i=start; i<number.length; i++){
            if(!check[i]){
                check[i] = 1;
                DFS(i,sum+number[i],cnt+1);
                check[i] = 0;
            }
        }
    }
    
    DFS(0,0,0);
    return answer
}

File: 프로그래머스/1/131705. 삼총사/삼총사.java


class Solution {
    
    static int answer = 0;
    static int[] number;
    
    public int solution(int[] number) {
        
        this.number = number;
        
        dfs(0,0,0);
        
        return answer;
    }
    
    private static void dfs(int index, int depth, int sum) {
        if (depth == 3) {
            if (sum == 0) {
                answer++;
            }
            return;
        }

        for (int i = index; i < number.length; i++) {
            dfs(i+1,depth + 1, sum + number[i]);
        }
    }
}